lib/phoenix_live_view/async_result.ex

```diff
@@ -0,0 +1,501 @@
+    <:canceled :let={_reason}>loading canceled</:canceled>
+    <%= org.name %>
+  <AsyncResult.with_state>
+  ```
+  """
+  attr :assign, :any, required: true
+  slot :loading
+
+  # TODO decide if we want an canceled slot
+  slot :canceled
+
+  # TODO decide if we want an empty slot
+  slot :empty,
+    doc:
+      "rendered when the result is loaded and is either nil or an empty list. Receives the result as a :let."
+
+  slot :failed,
+    doc:
+      "rendered when an error or exit is caught or assign_async returns `{:error, reason}`. Receives the error as a :let."
+
+  def with_state(assigns) do
+    case assigns.assign do
+      %AsyncResult{state: state, ok?: once_ok?, result: result} when state == :ok or once_ok? ->
+        if assigns.empty != [] && result in [nil, []] do
+          ~H|<%= render_slot(@empty, @assign.result) %>|
+        else
+          ~H|<%= render_slot(@inner_block, @assign.result) %>|
+        end
+
+      %AsyncResult{state: :loading} ->
+        ~H|<%= render_slot(@loading) %>|
+
+      %AsyncResult{state: {:error, {:canceled, reason}}} ->
+        if assigns.canceled != [] do
+          assigns = Phoenix.Component.assign(assigns, reason: reason)
+          ~H|<%= render_slot(@canceled, @reason) %>|
+        else
+          ~H|<%= render_slot(@failed, @assign.state) %>|
+        end
+
+      %AsyncResult{state: {kind, _reason}} when kind in [:error, :exit] ->
+        ~H|<%= render_slot(@failed, @assign.state) %>|
+    end
+  end
+
+  @doc """
+  Assigns keys asynchronously.
+
+  The task is linked to the caller and errors are wrapped.
+  Each key passed to `assign_async/3` will be assigned to
+  an `%AsyncResult{}` struct holding the status of the operation
+  and the result when completed.
+  """
+  def assign_async(%Socket{} = socket, key_or_keys, func)
+      when (is_atom(key_or_keys) or is_list(key_or_keys)) and
+             is_function(func, 0) do
+    keys = List.wrap(key_or_keys)
+
+    # verifies result inside task
+    wrapped_func = fn ->
+      case func.() do
+        {:ok, %{} = assigns} ->
+          if Map.keys(assigns) -- keys == [] do
+            {:ok, assigns}
+          else
+            raise ArgumentError, """
+            expected assign_async to return map of assigns for all keys
+            in #{inspect(keys)}, but got: #{inspect(assigns)}
+            """
+          end
+
+        {:error, reason} ->
+          {:error, reason}
+
+        other ->
+          raise ArgumentError, """
+          expected assign_async to return {:ok, map} of
+          assigns for #{inspect(keys)} or {:error, reason}, got: #{inspect(other)}
+          """
+      end
+    end
+
+    keys
+    |> Enum.reduce(socket, fn key, acc ->
+      async_result =
+        case acc.assigns do
+          %{^key => %AsyncResult{ok?: true} = existing} -> existing
+          %{} -> AsyncResult.new(key, keys)
+        end
+
+      Phoenix.Component.assign(acc, key, async_result)
+    end)
```

Comment:
Nitpick: something like this should be more efficient as there is less unwrapping around the socket:

```suggestion
    |> Enum.map(fn key, acc ->
        case acc.assigns do
          %{^key => %AsyncResult{ok?: true} = existing} -> {key, existing}
          %{} -> {key, AsyncResult.new(key, keys)}
        end
    )
    |> then(&Phoenix.Component.assign(socket, &1))
    
    
```